source/client_channel_handler.c

```diff
@@ -36,21 +36,50 @@ static int s_packet_handler_default(
     return aws_raise_error(AWS_ERROR_MQTT_INVALID_PACKET_TYPE);
 }
 
+static void s_schedule_ping_get_outbound_delta_data(
+    struct aws_mqtt_client_connection *connection,
+    uint64_t *result_now,
+    uint64_t *result_outbound_delta,
+    uint64_t *result_ping_time_ns) {
+    uint64_t now = 0;
+    aws_channel_current_clock_time(connection->slot->channel, &now);
+    uint64_t outbound_delta = now - connection->last_outbound_socket_write_time;
+    uint64_t ping_time_ns =
+        aws_timestamp_convert(connection->keep_alive_time_secs, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL);
+
+    if (result_now != NULL) {
+        *result_now = now;
+    }
+    if (result_outbound_delta != NULL) {
+        *result_outbound_delta = outbound_delta;
+    }
+    if (result_ping_time_ns != NULL) {
+        *result_ping_time_ns = ping_time_ns;
+    }
+}
+
 static void s_on_time_to_ping(struct aws_channel_task *channel_task, void *arg, enum aws_task_status status);
 static void s_schedule_ping(struct aws_mqtt_client_connection *connection) {
     aws_channel_task_init(&connection->ping_task, s_on_time_to_ping, connection, "mqtt_ping");
 
     uint64_t now = 0;
-    aws_channel_current_clock_time(connection->slot->channel, &now);
+    uint64_t outbound_delta = 0;
+    uint64_t ping_time_ns = 0;
+    s_schedule_ping_get_outbound_delta_data(connection, &now, &outbound_delta, &ping_time_ns);
+
     AWS_LOGF_TRACE(
-        AWS_LS_MQTT_CLIENT, "id=%p: Scheduling PING. current timestamp is %" PRIu64, (void *)connection, now);
+        AWS_LS_MQTT_CLIENT, "id=%p: Scheduling PING task. current timestamp is %" PRIu64, (void *)connection, now);
 
-    uint64_t schedule_time =
-        now + aws_timestamp_convert(connection->keep_alive_time_secs, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL);
+    uint64_t schedule_time = 0;
+    if (outbound_delta < ping_time_ns) {
```

Comment:
This is kind of weird and I think it's related to the fact that you're skipping updating the last outbound write time on pings.  I haven't mentally walked through things in detail, but I feel like pings shouldn't be treated especially and in doing so, it will simplify these time calculations which are not what I'd expect.  

I would expect this function to work properly no matter when it's called, but as it stands that's  not the case.  It seems coupled to a hidden "this function only gets called right after I've sent a ping that didn't update the last write time"

I can't tell if it's correct or not, but I can tell it would be unnecessarily tough to maintain after the fact.

Actually it doesn't seem correct.  Here's the basic scenario I'm blocked on:

Assume 30 second keep alive

T(0) - Connect and schedule ping
T(29) - Publish something, last outbound is updated
T(30) - Ping task goes off.  s_on_time_to_ping gets called, but we skip sending it (else clause below) and just call s_schedule_ping.  In s_schedule_ping, outbound_delta is 1, so schedule time is 30 + 30 + 1 = 61 when it should be 59.  It only gets worse the closer the publish is in time to the previous ping/connect timepoint.

Reply:
I fixed this. Now the PING packet is not treated specially and it works as expected. Now it marks the time regardless of whether it is a PING packet or another packet. I did have to add a slight delta check in the condition that checks to see if the time between the last write is less than the given keep alive, as a packet is not perfectly written exactly to the socket at the exact time (so we have to account for it). Now the code will account for 0.001 second delta accordingly.

I also fixed the reconnect timing so it should be as you mentioned, where it schedule at 59 rather than 61. I did some tests and I'm fairly confident that it functions like you said, where the schedule time will be 59 instead of 60 in the basic scenario given.

Reply:
However, I probably should do additional testing and make a test that basically is the scenario you mentioned above (with shorter times) to ensure that it is correctly working.

Reply:
Tests added! Now I am confident it is working correctly (even if in the second test I couldn't quite get the test suite to work how I wanted). I have checked the logs and confirmed that it should now work as expected in the basic scenario above.